proposed/logging-generator.md

```diff
@@ -0,0 +1,489 @@
+{
+    public static CustomLoggerParamsBuilder Create(int baseLength, int formatHoleCount, ILogger logger, LogLevel logLevel, out bool builderIsValid)
+    {
+        if (!logger.IsEnabled(logLevel))
+        {
+            builderIsValid = false;
+            return default;
+        }
+
+        builderIsValid = true;
+        return new CustomLoggerParamsBuilder(baseLength, formatHoleCount, logLevel);
+    }
+
+    private LogLevel _logLevelEnabled;
+
+    // returns the number of holes
+    public int Count => ... // depends on the type used to store key value pairs
+
+    public static readonly Func<CustomLoggerParamsBuilder, Exception?, string> FormatCallback = (builder, exception) => builder.ToString();
+
+    private CustomLoggerParamsBuilder(int baseLength, int formatHoleCount, LogLevel logLevelEnabled)
+    {
+        // Initialization logic
+        _logLevelEnabled = logLevelEnabled;
+    }
+
+    public bool TryFormatBaseString(string s)
+    {
+        // Store and format part as required
+        return true;
+    }
+
+    public bool TryFormatInterpolationHole<T>(T t)
+    {
+        // Store and format part as required
+        // Store name hole and value as KeyValuePair<string, object?>
+        return true;
+    }
+
+    public KeyValuePair<string, object?> this[int index]
+    {
+        get => ... // depends on the type used to store key value pairs
+    }
+
+    public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+    {
+        for (int i = 0; i < Count; ++i)
+        {
+            yield return this[i];
+        }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+}
+```
+
+### Q & A - The builder approach:
+
+- Question: Can I use a custom string interpolation builder in an async context?
+
+```csharp
+public async Task Method()
+{
+    // Cannot use MyStringBuilder if it's a ref struct
+    var key1 = "value1";
+    var key2 = "value2";
+    logger.LogInformation($"Some text: {key1}, {key2}");
+    // ...
+    await Method2();
+}
+```
+
+Answer: Yes, as long as the builder is not a ref struct.
+
+- Question: Is duck typing supported for this string interpolated builder language feature?
+
+Answer: Yes, as long as `TryFormatX` APIs are defined on a new builder type, then this feature would be enabled. 
+
+### Sample-based comparison of two approaches
+
+1. (+1 for the C# 10 Language Feature) _With the interpolated string builder, callers don't need to guard the call sites if any computation is necessary to produce the arguments to the logging. Whereas with the source generator, the user call site would need to get wrapped around `IsEnabled` checks._
+
+For example with the source generator:
+
+```csharp
+[LoggerMessage(2, LogLevel.Trace, "Certificates: `{matchingCertificates}`")]
+public static partial void DecribeFoundCertificates(this ILogger logger, IEnumerable<X509Certificate2> matchingCertificates);
```

Comment:
Can you write this with the source generator?  What translates the IEnumerable into the resulting string?

Reply:
Yes, could lead to:

```csharp  
[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "1.0.0.0")]
private static readonly global::System.Action<global::Microsoft.Extensions.Logging.ILogger, System.Collections.Generic.IEnumerable<string>, global::System.Exception?> _DecribeFoundCertificatesCallback =
    global::Microsoft.Extensions.Logging.LoggerMessage.Define<System.Collections.Generic.IEnumerable<string>>(global::Microsoft.Extensions.Logging.LogLevel.Trace, new global::Microsoft.Extensions.Logging.EventId(2, nameof(DecribeFoundCertificates)), "Certificates: `{matchingCertificates}`", true); 

[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "1.0.0.0")]
public static partial void DecribeFoundCertificates(this Microsoft.Extensions.Logging.ILogger logger, System.Collections.Generic.IEnumerable<string> matchingCertificates)
{
    if (logger.IsEnabled(global::Microsoft.Extensions.Logging.LogLevel.Trace))
    {
        _DecribeFoundCertificatesCallback(logger, matchingCertificates, null);
    }
}
```

The underlying `LoggerMessage.Define<IEnumerable<string>>(..)` call translates `IEnumerable` to string, (using an internal `LogValues` type as TState: [link to implementation](https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs,241))

Reply:
But if the dev actually wanted to control the string rather than it just concatenating ToString()s of the enumerated items, e.g.
https://github.com/dotnet/aspnetcore/blob/277cc564ce5a76c6613f29dd92c0168e06b41b1c/src/Shared/CertificateGeneration/CertificateManager.cs#L104
presumably they would instead need to write something like this:
```C#
public static void DecribeFoundCertificates(this ILogger logger, IEnumerable<X509Certificate2> matchingCertificates)
{
    if (logger.IsEnabled(LogLevel.Trace)
    {
        DecribeFoundCertificates(ToCertificateDescription(matchingCertificates));
    }
}

[LoggerMessage(2, LogLevel.Trace, "Certificates: `{matchingCertificates}`")]
private static partial void DecribeFoundCertificates(this ILogger logger, string matchingCertificates);
```
?

Reply:
That's correct. That's the pattern we use today

Reply:
Yes sure, that looks like a valid sample code, when developer wants to control the string itself.

Reply:
I'll update the document sample using your snippet above